src/hotspot/share/gc/g1/g1EvacFailureRegions.inline.hpp

```diff
@@ -27,15 +27,18 @@
 
 #include "gc/g1/g1EvacFailureRegions.hpp"
 #include "runtime/atomic.hpp"
-#include "utilities/bitMap.inline.hpp"
 
 bool G1EvacFailureRegions::record(uint region_idx) {
-  assert(region_idx < _max_regions, "must be");
   bool success = _regions_failed_evacuation.par_set_bit(region_idx,
                                                         memory_order_relaxed);
   if (success) {
     size_t offset = Atomic::fetch_and_add(&_evac_failure_regions_cur_length, 1u);
     _evac_failure_regions[offset] = region_idx;
+
+    G1CollectedHeap* g1h = G1CollectedHeap::heap();
+    HeapRegion* hr = g1h->region_at(region_idx);
+    G1CollectorState* state = g1h->collector_state();
+    hr->note_self_forwarding_removal_start(state->in_concurrent_start_gc());
```

Comment:
It's unclear to me why it's called *here* -- we are still in the evacuation phase, self-forwarding-removal occurs in post-evacuation, isn't it?

Reply:
> void HeapRegion::note_self_forwarding_removal_end_par(size_t garbage_bytes) {
  Atomic::store(&_live_bytes, region_size - _garbage_bytes, ...);
  Atomic::add(&_garbage_bytes, garbage_bytes, memory_order_relaxed);
}

This comment seems to be misplaced - the suggested method seems unrelated to `mark_evac_failure_object`. Also, there is no `_live_bytes` member in `HeapRegion`; probably `_marked_bytes` has been meant, but that has been removed in [JDK-8290357](https://bugs.openjdk.org/browse/JDK-8290357) because it's not required to actually store that value. Live bytes is `top() - garbage_bytes()`.

> It's unclear to me why it's [`note_self_forwarding_removal_start`] called here -- we are still in the evacuation phase, self-forwarding-removal occurs in post-evacuation, isn't it?

`note_self_forwarding_removal_start` prepares for self forwarding removal which is unavoidable at that point for that region. Previously the call to this method during self forward removal was easily possible because one thread only every worked on a single region.
The original code had an extra phase that iterated over all regions just for that, which has been removed [here](https://github.com/openjdk/jdk/pull/7047/commits/1003c77cfebc3a72f409d2ebff41b8a74c3e4495) I think, and this call placed here because of lack of other really good options (that are more complicated).

I will rename the methods/comments.